From 78af9038cd42103eae350106406eeed90f78307b Mon Sep 17 00:00:00 2001
From: Vasyl Gello <vasek.gello@gmail.com>
Date: Tue, 7 May 2019 07:58:21 +0300
Subject: [PATCH] Merge android-9.0.0_r36 into android-6.0.1

Squashed commit of the following:

commit 5afe728e4b3cd184f9fb99f58c853e9e8c1ef692
Merge: f8830371 f04f2b72
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Wed Mar 13 18:05:42 2019 +0000

    Merge cherrypicks of [6716922, 6716923, 6716413, 6717023, 6717024, 6716716, 6715859, 6717160, 6717161, 6717162, 6717163, 6716295, 6717141, 6717181, 6717183, 6717184, 6717185, 6714937, 6717028, 6716717, 6716927, 6717200, 6717029, 6717030, 6717031, 6717032, 6717033, 6716928, 6717034, 6717035, 6716929, 6717201, 6716930, 6712377, 6712378, 6716643, 6717164, 6712379] into pi-qpr3-release

    Change-Id: I29502c4e125493a9bfaba45e4ab2a35c8d97c5d4

commit f04f2b72c6656659274eacb51dd7558e19914f89
Author: Bryan Ferris <bferris@google.com>
Date:   Tue Mar 5 16:19:48 2019 -0800

    Fix type confusion in libpac

    From the upstream patch
    (https://chromium.googlesource.com/v8/v8.git/+/55a98076827edac8eba775f8025df3749bcd8367%5E%21/#F0):

    """
    Fix regexp fast path in MaybeCallFunctionAtSymbol

    The regexp fast path in MaybeCallFunctionAtSymbol had an issue in which
    we'd call ToString after checking that the given {object} was a fast
    regexp and deciding to take the fast path. This is invalid since
    ToString() can call into user-controlled JS and may mutate {object}.

    There's no way to place the ToString call correctly in this instance:
    1 before BranchIfFastRegExp, it's a spec violation if we end up on the
      slow regexp path;
    2 the problem with the current location is already described above;
    3 and we can't place it into the fast-path regexp builtin (e.g.
      RegExpReplace) either due to the same reasons as 1.

    The solution in this CL is to restrict the fast path to string
    arguments only, i.e. cases where ToString would be a nop and can safely
    be skipped.
    """

    Bug: 117556606

    Test: /data/nativetest/proxy_resolver_v8_unittest/proxy_resolver_v8_unittest

    Test: gts-tradefed run gts --test \
      com.google.android.gts.devicepolicy.DeviceOwnerTest#testProxyPacProxyTest \
      --module GtsGmscoreHostTestCases

    Test: PoC from bug report

    Merged-In: I2e02d994f107e64e4f465b4d8a02d4159a95240e

    Change-Id: Ifb58de2b3c547c442f1ad69e0bca0fa934d1f728
    (cherry picked from commit ce91afbb1b8ed1c0bbde11609be1f93e4bbfa461)

commit 9817e77d44b65abce2aba7f4d6c6372ed09a4a52
Author: Bryan Ferris <bferris@google.com>
Date:   Tue Mar 5 14:38:25 2019 -0800

    [RESTRICT AUTOMERGE] Fix OOB Access in libpac

    From upstream commit
    (https://chromium.googlesource.com/v8/v8.git/+/389539ca30406101971745820d3105e7f39bb772%5E%21/#F0):

    """
    Fix type of String#indexOf and String#lastIndexOf.

    The Typer put the wrong type on String#index and String#lastIndexOf
    builtins, with an off by one on the upper bound.
    """

    Bug: 117554758

    Test: /data/nativetest/proxy_resolver_v8_unittest/proxy_resolver_v8_unittest

    Test: gts-tradefed run gts --test \
      com.google.android.gts.devicepolicy.DeviceOwnerTest#testProxyPacProxyTest \
      --module GtsGmscoreHostTestCases

    Test: PoC from bug report

    Merged-In: I2e02d994f107e64e4f465b4d8a02d4159a95240e

    Change-Id: I6230c1b764cacf9349317ad5870d7a679600d7d5
    (cherry picked from commit 32ab7b6d784f24161d312f67a505c71f7c34df40)

commit e3b988edf9cdfb318ba7aa1d8e3c0ddea53892de
Author: Bryan Ferris <bferris@google.com>
Date:   Tue Mar 5 18:06:51 2019 -0800

    Fix Integer Overflow in libpac

    From the upstream patch
    (https://chromium.googlesource.com/v8/v8.git/+/7b27040e66c8a83006aebc90fe97d21bb42156a7%5E%21/#F0):

    """
    Harden JSFunction::CalculateInstanceSizeHelper(...)
    """

    Bug: 117556220

    Test: /data/local/nativetest/proxy_resolver_v8_unittest/proxy_resolver_v8_unittest

    Test: gts-tradefed run gts --test \
      com.google.android.gts.devicepolicy.DeviceOwnerTest#testProxyPacProxyTest \
      --module GtsGmscoreHostTestCases

    Test: PoC from bug report

    Merged-In: I2e02d994f107e64e4f465b4d8a02d4159a95240e

    Change-Id: I72321236f4d4cf8da993addc3ef7a1dc018c434b
    (cherry picked from commit cc15c8f69c835e5c1fcfd4992cb21adbda41bd03)

commit 8279c47df01ac6dc28d67c411dfa36c61a068c62
Author: Bryan Ferris <bferris@google.com>
Date:   Tue Mar 5 17:18:50 2019 -0800

    Fix type confusion in libpac

    From the upstream patch
    (https://chromium.googlesource.com/v8/v8/+/ea55b873f2ed8336604540a532cbd460eeb66430%5E%21/#F0):

    """
    Don't generate elements kind transitions from stable maps.

    IC system does its best to properly mark stable transition source maps
    as unstable (see https://chromium-review.googlesource.com/483442)
    however an already recorded map can be deprecated later and the
    optimizing compiler may try to generate an elements kind transition
    from the updated version of deprecated map which can "become" stable
    again.
    """

    Bug: 117607414

    Test: /data/nativetest/proxy_resolver_v8_unittest/proxy_resolver_v8_unittest

    Test: gts-tradefed run gts --test \
      com.google.android.gts.devicepolicy.DeviceOwnerTest#testProxyPacProxyTest \
      --module GtsGmscoreHostTestCases

    Test: PoC from bug report

    Merged-In: I2e02d994f107e64e4f465b4d8a02d4159a95240e

    Change-Id: I12c501bffd190e20d4a45a4256a403c5343350eb
    (cherry picked from commit 97832faf9d62ce9d6ca0639eacac0a930e51df8a)

commit a84c4b763ed287f762e4112c73d9fb357605b02c
Author: Bryan Ferris <bferris@google.com>
Date:   Tue Mar 5 15:05:33 2019 -0800

    Fix OOB read in libpac ast-numbering.cc

    From the upstream patch
    (https://chromium.googlesource.com/v8/v8.git/+/fbf974e2af0f62b075e62701581099211f83adb6%5E%21/#F0):

    """
    Add missing early-bailouts in ast traversal visitors
    """

    Bug: 117555811

    Test: adb shell /data/nativetest/proxy_resolver_v8_unittest/proxy_resolver_v8_unittest

    Test: gts-tradefed run gts --test \
      com.google.android.gts.devicepolicy.DeviceOwnerTest#testProxyPacProxyTest \
      --module GtsGmscoreHostTestCases

    Test: PoC from bug report

    Merged-In: I2e02d994f107e64e4f465b4d8a02d4159a95240e

    Change-Id: Ia62b5ea79d79400166b7e238be1933c55730612b
    (cherry picked from commit 8cc1440e71299ca16640b4371adf378d0a61983f)

commit f88303717d5ed67e6ba13c7177ebd3aada958caf
Merge: 095675d8 32f6b6a6
Author: Pavel Grafov <pgrafov@google.com>
Date:   Thu Aug 9 06:29:04 2018 -0700

    [automerger skipped] Merge "Backport: Fix Object.entries/values with changing elements" into oc-dev am: 535a1a144c am: 79030d25fa  -s ours
    am: 32f6b6a65e  -s ours

    Change-Id: Ib85415140fbe32328f5dac069ef5d7a2cab3596b

commit 32f6b6a65e5174f4de76231f19c0d1566fecdbe5
Merge: eb77944d 79030d25
Author: Pavel Grafov <pgrafov@google.com>
Date:   Thu Aug 9 06:23:22 2018 -0700

    [automerger skipped] Merge "Backport: Fix Object.entries/values with changing elements" into oc-dev am: 535a1a144c
    am: 79030d25fa  -s ours

    Change-Id: I53657a39b565099ea667f7d7c8733a5c21a4ecc2

commit 79030d25fa5e197808f056082c9a7b914d794e88
Merge: d544c38c 535a1a14
Author: Pavel Grafov <pgrafov@google.com>
Date:   Thu Aug 9 06:18:40 2018 -0700

    Merge "Backport: Fix Object.entries/values with changing elements" into oc-dev
    am: 535a1a144c

    Change-Id: I0bf35383206c93816a005498ace6aea0a2442189

commit 535a1a144c95bf90eebc2ecdd01cb2c99d2a89dd
Merge: 538849e2 275d7152
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Thu Aug 9 13:08:44 2018 +0000

    Merge "Backport: Fix Object.entries/values with changing elements" into oc-dev

commit 095675d8d20415a7e08cb7876dd47259308005ce
Merge: 53d4be7c eb77944d
Author: Pavel Grafov <pgrafov@google.com>
Date:   Thu Aug 9 03:29:14 2018 -0700

    [automerger skipped] [automerger] DO NOT MERGE Backport: Fix Object.entries/values with changing elements am: ef942a244d am: 9a8e7d8386 am: c11715fe14 am: 2aba1639d8 skipped: 8892439df8 am: 538849e226 am: d544c38cf1
    am: eb77944d3d  -s ours

    Change-Id: I51c8b6ec0cf113194b4555b8dca3d34fa5adc9ef

commit eb77944d3da3f9aa2f270d9d4825f75fe2b4052c
Merge: d8be9a10 d544c38c
Author: Pavel Grafov <pgrafov@google.com>
Date:   Thu Aug 9 03:25:52 2018 -0700

    [automerger] DO NOT MERGE Backport: Fix Object.entries/values with changing elements am: ef942a244d am: 9a8e7d8386 am: c11715fe14 am: 2aba1639d8 skipped: 8892439df8 am: 538849e226
    am: d544c38cf1

    Change-Id: I36461401a8b7c55530c885eff4bfb459364f5953

commit 53d4be7c7a9a09122ae18606f37f44aad1b73e02
Merge: c86a7f89 d8be9a10
Author: Pavel Grafov <pgrafov@google.com>
Date:   Thu Aug 9 03:24:46 2018 -0700

    Backport: Fix Object.entries/values with changing elements
    am: d8be9a1028

    Change-Id: I0085359cc58be0443066961a60304e659d5cb04a

commit d544c38cf10645da7f67e5dbfce9c2f5fb3496a8
Merge: c06d79b8 538849e2
Author: Pavel Grafov <pgrafov@google.com>
Date:   Thu Aug 9 03:19:48 2018 -0700

    [automerger] DO NOT MERGE Backport: Fix Object.entries/values with changing elements am: ef942a244d am: 9a8e7d8386 am: c11715fe14 am: 2aba1639d8 skipped: 8892439df8
    am: 538849e226

    Change-Id: I42781136876ce1b23fd00630f4fd5ed6e08c11f9

commit 275d7152a6029056732896bd45fd05a243b0bf4f
Author: Pavel Grafov <pgrafov@google.com>
Date:   Mon Jul 30 14:22:32 2018 +0100

    Backport: Fix Object.entries/values with changing elements

    Bug: 111274046
    Test: m -j proxy_resolver_v8_unittest && adb sync && adb shell \
    /data/nativetest64/proxy_resolver_v8_unittest/proxy_resolver_v8_unittest
    Merged-In: I705fc512cc5837e9364ed187559cc75d079aa5cb
    Change-Id: I58c84a46949b79db44b1bb4f17794b54ebb9a301

commit d8be9a10287afed07705ac8af027d6a46d4def99
Author: Pavel Grafov <pgrafov@google.com>
Date:   Thu Aug 2 20:18:07 2018 +0100

    Backport: Fix Object.entries/values with changing elements

    Bug: 111274046
    Test: m -j proxy_resolver_v8_unittest && adb sync && adb shell \
    /data/nativetest64/proxy_resolver_v8_unittest/proxy_resolver_v8_unittest
    Change-Id: I705fc512cc5837e9364ed187559cc75d079aa5cb

commit 538849e2263b3d170aae09b582315b9180cdeb0a
Merge: 24bb22a0 8892439d
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu Aug 2 21:19:59 2018 +0000

    [automerger] DO NOT MERGE Backport: Fix Object.entries/values with changing elements am: ef942a244d am: 9a8e7d8386 am: c11715fe14 am: 2aba1639d8 skipped: 8892439df8

    Change-Id: I94ae46c5b8772a94a7a14496def1c15ed97b0b32

commit 8892439df8aad1f5ef9218b1f80608d8d0ab1c41
Merge: 1636f6eb 2aba1639
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu Aug 2 21:19:56 2018 +0000

    [automerger] DO NOT MERGE Backport: Fix Object.entries/values with changing elements am: ef942a244d am: 9a8e7d8386 am: c11715fe14 am: 2aba1639d8

    Change-Id: Ida1defa4ae4387de9cc8901db2e12aeb90a6765e

commit 2aba1639d80ab73b58616e1e73555fec8a3b6764
Merge: 5556d694 c11715fe
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu Aug 2 21:19:54 2018 +0000

    [automerger] DO NOT MERGE Backport: Fix Object.entries/values with changing elements am: ef942a244d am: 9a8e7d8386 am: c11715fe14

    Change-Id: I8a3e2d37639624cc9f7c4b0458bcb91ea3fab8d9

commit c11715fe14b697175383ecbf78d0595c62d82d8c
Merge: 8776b2b9 9a8e7d83
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu Aug 2 21:19:51 2018 +0000

    [automerger] DO NOT MERGE Backport: Fix Object.entries/values with changing elements am: ef942a244d am: 9a8e7d8386

    Change-Id: I19998a4abfbe798d8790dc6cd43004a08768c2ca

commit 9a8e7d83866b115a53810a5880fa289909825957
Merge: bb7c3e3b ef942a24
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu Aug 2 21:19:49 2018 +0000

    [automerger] DO NOT MERGE Backport: Fix Object.entries/values with changing elements am: ef942a244d

    Change-Id: Id9677739af925d8babaf458f88326a874f5589d2

commit ef942a244dd8b1c9b3707430b826c90ef79a897a
Author: Pavel Grafov <pgrafov@google.com>
Date:   Thu Aug 2 20:18:07 2018 +0100

    DO NOT MERGE Backport: Fix Object.entries/values with changing elements

    Bug: 111274046
    Test: m -j proxy_resolver_v8_unittest && adb sync && adb shell \
    /data/nativetest64/proxy_resolver_v8_unittest/proxy_resolver_v8_unittest
    Change-Id: Ie0b83ceeab4e892a3eb0b9ddd0a9b9ba20e884d5

commit c86a7f894c3c0abbe035ce0699b084b9c258a1a9
Merge: d763d99f 8e7b4a2b
Author: Rubin Xu <rubinxu@google.com>
Date:   Thu May 3 15:39:08 2018 -0700

    [automerger skipped] [automerger] [DO NOT MERGE] [runtime] Fix Array.prototype.concat with complex @@species am: d8c6189dcb am: 05a61246f1 am: 0e4e8538f1 skipped: 8879f3a573 am: 691e1f280f am: bb7c3e3b80 am: 8776b2b9f1 am: 5556d6945b skipped: 1636f6ebda am: 24bb22a06c  -s ours am: c06d79b843  -s ours
    am: 8e7b4a2b92  -s ours

    Change-Id: I3c4eb5b5b299a855d282ff8ecf31ca1104a7715b

commit 8e7b4a2b92da3be5a0f36a64825e6081919a2144
Merge: 16015780 c06d79b8
Author: Rubin Xu <rubinxu@google.com>
Date:   Thu May 3 15:34:41 2018 -0700

    [automerger skipped] [automerger] [DO NOT MERGE] [runtime] Fix Array.prototype.concat with complex @@species am: d8c6189dcb am: 05a61246f1 am: 0e4e8538f1 skipped: 8879f3a573 am: 691e1f280f am: bb7c3e3b80 am: 8776b2b9f1 am: 5556d6945b skipped: 1636f6ebda am: 24bb22a06c  -s ours
    am: c06d79b843  -s ours

    Change-Id: I5cf5960ac860497877d8698ce5bab03ac3bacda8

commit c06d79b8439638f82ea34fea46a49418738e2b0c
Merge: c7ffcd07 24bb22a0
Author: Rubin Xu <rubinxu@google.com>
Date:   Thu May 3 15:29:39 2018 -0700

    [automerger skipped] [automerger] [DO NOT MERGE] [runtime] Fix Array.prototype.concat with complex @@species am: d8c6189dcb am: 05a61246f1 am: 0e4e8538f1 skipped: 8879f3a573 am: 691e1f280f am: bb7c3e3b80 am: 8776b2b9f1 am: 5556d6945b skipped: 1636f6ebda
    am: 24bb22a06c  -s ours

    Change-Id: Iebb9008b9d833052cc4d073b4f26e5269eb19e21

commit d763d99fadc7149e0eefea8e66b1192dcbe369d4
Merge: 7431855e 16015780
Author: Stephen Li <stephenli@google.com>
Date:   Thu May 3 15:12:02 2018 -0700

    [automerger skipped] Merge commit '63e855d3cbb9a29677c7f5381b5a13b3bce4de5b' into nyc-mr2-dev skipped: 383e4ce692 am: a8cb31350e am: c7ffcd0712
    am: 16015780da

    Change-Id: I976ae1c233c2df9b09844af16a6101f2119ac52d

commit 16015780da31a9a2240c68dedb49569d6b1110dd
Merge: a938e3cd c7ffcd07
Author: Stephen Li <stephenli@google.com>
Date:   Thu May 3 14:53:25 2018 -0700

    [automerger skipped] Merge commit '63e855d3cbb9a29677c7f5381b5a13b3bce4de5b' into nyc-mr2-dev skipped: 383e4ce692 am: a8cb31350e
    am: c7ffcd0712

    Change-Id: Ia8b88af4d194917a3a5fede6b885bb4d31c1cd92

commit 24bb22a06c36d0e7ab3b0cb8be800c90fa7110e9
Merge: a8cb3135 1636f6eb
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu May 3 21:28:50 2018 +0000

    [automerger] [DO NOT MERGE] [runtime] Fix Array.prototype.concat with complex @@species am: d8c6189dcb am: 05a61246f1 am: 0e4e8538f1 skipped: 8879f3a573 am: 691e1f280f am: bb7c3e3b80 am: 8776b2b9f1 am: 5556d6945b skipped: 1636f6ebda

    Change-Id: I8195adcabd509f7e19ae11ed04b4e22812790ec5

commit 1636f6ebdad2febf2a0d1b35f6cddde8cd9bcc24
Merge: 383e4ce6 5556d694
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu May 3 21:28:47 2018 +0000

    [automerger] [DO NOT MERGE] [runtime] Fix Array.prototype.concat with complex @@species am: d8c6189dcb am: 05a61246f1 am: 0e4e8538f1 skipped: 8879f3a573 am: 691e1f280f am: bb7c3e3b80 am: 8776b2b9f1 am: 5556d6945b

    Change-Id: I82f1a94305294fbaa96b3b3dd7e684df11c9d2d3

commit 5556d6945b5f13d145b7ee3f64a43276bfd532ba
Merge: 63e855d3 8776b2b9
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu May 3 21:28:45 2018 +0000

    [automerger] [DO NOT MERGE] [runtime] Fix Array.prototype.concat with complex @@species am: d8c6189dcb am: 05a61246f1 am: 0e4e8538f1 skipped: 8879f3a573 am: 691e1f280f am: bb7c3e3b80 am: 8776b2b9f1

    Change-Id: Ib284a3469af678c0beb4b163479a51a5793b2a4e

commit 8776b2b9f1fd1188861373896dbdbcd852c1627d
Merge: 5b7e7d54 bb7c3e3b
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu May 3 21:28:43 2018 +0000

    [automerger] [DO NOT MERGE] [runtime] Fix Array.prototype.concat with complex @@species am: d8c6189dcb am: 05a61246f1 am: 0e4e8538f1 skipped: 8879f3a573 am: 691e1f280f am: bb7c3e3b80

    Change-Id: I231425dfbc2d4491a4d7f58f00d14ff3dab8dd89

commit bb7c3e3b80168ccc2f84b43790092443da973929
Merge: 9ff032da 691e1f28
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu May 3 21:28:41 2018 +0000

    [automerger] [DO NOT MERGE] [runtime] Fix Array.prototype.concat with complex @@species am: d8c6189dcb am: 05a61246f1 am: 0e4e8538f1 skipped: 8879f3a573 am: 691e1f280f

    Change-Id: I0a8000389bd85626c19cbe5d9d91ab6088d50b0b

commit 691e1f280f1f35d36493717171e83019bea83f93
Merge: 1f31521f 8879f3a5
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu May 3 21:28:39 2018 +0000

    [automerger] [DO NOT MERGE] [runtime] Fix Array.prototype.concat with complex @@species am: d8c6189dcb am: 05a61246f1 am: 0e4e8538f1 skipped: 8879f3a573

    Change-Id: I042bf14b600ded6208edc5eb4be2a3d93516b2da

commit 8879f3a573ab1293a77b8cbcd06a9ce9979e56a9
Merge: 749f88d6 0e4e8538
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu May 3 21:23:13 2018 +0000

    [automerger] [DO NOT MERGE] [runtime] Fix Array.prototype.concat with complex @@species am: d8c6189dcb am: 05a61246f1 am: 0e4e8538f1

    Change-Id: I5e12cc8e024a9c8047aa8b36db7aea29edbf9423

commit 0e4e8538f1a9027269a61d48a7f6d220866bcd55
Merge: 3a5398ae 05a61246
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu May 3 21:23:10 2018 +0000

    [automerger] [DO NOT MERGE] [runtime] Fix Array.prototype.concat with complex @@species am: d8c6189dcb am: 05a61246f1

    Change-Id: I8f8268c9233c3b5ad8f7a095857283db8ecdd94e

commit 05a61246f14ffde0d9554049a526294df6c1458e
Merge: 36e9ff16 d8c6189d
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu May 3 21:23:08 2018 +0000

    [automerger] [DO NOT MERGE] [runtime] Fix Array.prototype.concat with complex @@species am: d8c6189dcb

    Change-Id: If9b9de9c4e9498be59a38fb586483b6aec644fb2

commit d8c6189dcb76343d80f76907a1d728f8635bc20b
Author: Rubin Xu <rubinxu@google.com>
Date:   Fri Mar 2 10:09:17 2018 +0000

    [DO NOT MERGE] [runtime] Fix Array.prototype.concat with complex @@species

    Manually cherry-picked from https://codereview.chromium.org/2666543004/

    Bug: 38196219
    Test: POC no longer crashes PacProcessor
    Change-Id: I8a8ed1806c6a430f23ce20c6de8dc266a76c19ab
    Merged-In: Ifc1a9f3ab8cbc0928c58951db78b3f9ae15449e4

commit c7ffcd0712e453e8f631ec1fa5bca35b16e2f28c
Merge: d424a9e9 a8cb3135
Author: Stephen Li <stephenli@google.com>
Date:   Thu May 3 14:21:17 2018 -0700

    [automerger skipped] Merge commit '63e855d3cbb9a29677c7f5381b5a13b3bce4de5b' into nyc-mr2-dev skipped: 383e4ce692
    am: a8cb31350e

    Change-Id: I57cddcd86479a69ccf52c9b04cb521cd91293940

commit a8cb31350e5f737d8dbda29f115fe460d6d267ac
Merge: 922f3522 383e4ce6
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu May 3 21:11:28 2018 +0000

    [automerger skipped] Merge commit '63e855d3cbb9a29677c7f5381b5a13b3bce4de5b' into nyc-mr2-dev skipped: 383e4ce692

    Change-Id: Id6675d2743a1c8cdb8c280bab89d1e26cfd24607

commit 383e4ce6923fafc2ccda52b194bd78371e59e429
Merge: 5a825a8d 63e855d3
Author: Stephen Li <stephenli@google.com>
Date:   Thu May 3 13:07:47 2018 -0700

    Merge commit '63e855d3cbb9a29677c7f5381b5a13b3bce4de5b' into nyc-mr2-dev

    BUG:78591499
    Change-Id: Idc461b592c454d5c6e016df22b187b6db16c3588

commit 63e855d3cbb9a29677c7f5381b5a13b3bce4de5b
Merge: 8316140f 5b7e7d54
Author: Stephen Li <stephenli@google.com>
Date:   Thu May 3 13:04:59 2018 -0700

    Merge commit '5b7e7d5493583847f522d94a4d9fd76a95f077c4' into cw-f-dev

    BUG:78591499
    Change-Id: I6e4b4f3f54477ba1c45e715258e7db70fcae3770

commit 5b7e7d5493583847f522d94a4d9fd76a95f077c4
Merge: b51fda4b 9ff032da
Author: Stephen Li <stephenli@google.com>
Date:   Thu May 3 12:57:42 2018 -0700

    Merge commit '9ff032da6b6d9dda9ae2324c22ff4a55c33c9384' into nyc-mr1-dev

    BUG:78591499
    Change-Id: Icb52e78b02ce66c0bd6e402835661aa9b219b797

commit 9ff032da6b6d9dda9ae2324c22ff4a55c33c9384
Merge: 6ae655ae 1f31521f
Author: Stephen Li <stephenli@google.com>
Date:   Thu May 3 12:49:52 2018 -0700

    Manually merge commit '1f31521f4ba2a8200edaac92be63f0d1d2af0598' into nyc-dr1-dev

    BUG:78591499
    Change-Id: I8961331bc56cd0c3cecf2c87f6e7caba9b9297fe

commit 1f31521f4ba2a8200edaac92be63f0d1d2af0598
Author: Ben Murdoch <benm@google.com>
Date:   Tue Jan 3 16:15:18 2017 -0800

    Merge up to V8 5.8. DO NOT MERGE

    62ed631 Merge V8 5.8.283.32
    ce06aad Add a script to help automate V8 updates.
    bd6fedf Remove unncessary -no-integrated-as flag.
    1e06733 Add missing #include "unicode/ustring.h" statement.
    b3e7c2e Fix build break on Mac.
    c8c1d9e Merge V8 5.6.326.50
    f3b273f Merge V8 5.5.372.32
    01e4440 Fix warnings generated by new makefiles.
    f91f061 Merge V8 5.4.500.40
    8bff9c4 Remove submodule references
    0c7622c Use -O2 for d8 on Mips

    Test: Build and ran D8. Verified PAC processing manually.

    Bug: 38196219
    Merged-In: Ifc1a9f3ab8cbc0928c58951db78b3f9ae15449e4
    Change-Id: Ifc1a9f3ab8cbc0928c58951db78b3f9ae15449e4

commit 7431855e0638bff8222258d5a91c7e84e9a10959
Merge: 7caea81d 4b98b001
Author: Chih-hung Hsieh <chh@google.com>
Date:   Wed Feb 14 23:34:52 2018 +0000

    Merge "Suppress clang 7.0 unused-private-field warnings." am: b2a7095332 am: 2659f86310
    am: 4b98b0013b

    Change-Id: Ie0d2621036cc51b9a43da0e7fc537586d90589c7

commit 4b98b0013b4b5495928af2abd1b002c576aaccaa
Merge: 44ba3dc3 2659f863
Author: Chih-hung Hsieh <chh@google.com>
Date:   Wed Feb 14 23:29:17 2018 +0000

    Merge "Suppress clang 7.0 unused-private-field warnings." am: b2a7095332
    am: 2659f86310

    Change-Id: Ibf275967578f92123f9e85213f435833b4bfb909

commit 2659f86310932947705457abfa59842ebc0f4541
Merge: 20f89107 b2a70953
Author: Chih-hung Hsieh <chh@google.com>
Date:   Wed Feb 14 23:23:57 2018 +0000

    Merge "Suppress clang 7.0 unused-private-field warnings."
    am: b2a7095332

    Change-Id: I0aeabea2e2950f9a0f256d57bcc6c77cb7941e6f

commit b2a709533213472c32e1c3928d0a285e64ef5dde
Merge: 20f89107 8585a772
Author: Chih-hung Hsieh <chh@google.com>
Date:   Wed Feb 14 23:12:59 2018 +0000

    Merge "Suppress clang 7.0 unused-private-field warnings."

commit 8585a772abd7cd64cf9ef6543a899ebe35558621
Author: Chih-Hung Hsieh <chh@google.com>
Date:   Tue Jan 16 10:22:18 2018 -0800

    Suppress clang 7.0 unused-private-field warnings.

    Test: make checkbuild
    Change-Id: Iedac1d023bdd08a983454e627b1656d428f87c36

commit 7caea81d0d41a868fa340a9eb3e705a1ba16cce0
Merge: 5835d76a 44ba3dc3
Author: Xin Li <delphij@google.com>
Date:   Wed Nov 15 02:47:11 2017 +0000

    Merge remote-tracking branch 'goog/stage-aosp-master' into HEAD
    am: 44ba3dc33c

    Change-Id: I11d46bf36b7b5134bda22926a86e7daa10fb1313

commit 44ba3dc33c41542d27511ddd0a7f2b5efb00f3cb
Merge: 35a1770b 20f89107
Author: Xin Li <delphij@google.com>
Date:   Tue Nov 14 16:38:15 2017 -0800

    Merge remote-tracking branch 'goog/stage-aosp-master' into HEAD

    Change-Id: I09caa5b6663a1d48fdfee3da75edecceeef50826

commit 20f891076ec0c29410ac08639db63a3ef6080cab
Merge: 25eea941 a938e3cd
Author: Xin Li <delphij@google.com>
Date:   Mon Nov 13 14:35:31 2017 -0800

    Merge commit 'a938e3cdb29f70199b4dbf3a5438979f57eda327' into HEAD

    Change-Id: I5b32b805c52ad40d480be6cf5b8e793b681bee6c

commit 5835d76aea2d3473f8b97d81557f107ff0ac9c52
Merge: a938e3cd 35a1770b
Author: Chih-hung Hsieh <chh@google.com>
Date:   Thu Oct 19 21:39:39 2017 +0000

    Merge "Use -Werror in external/v8" am: 91ca7d58d4 am: 25eea94176 am: 1d75d7df90
    am: 35a1770b8c

    Change-Id: Idc4379065141bdb555b58f0e90e33d687806928c

commit 35a1770b8c41d33c5902a37563cb070462cbc068
Merge: a938e3cd 1d75d7df
Author: Chih-hung Hsieh <chh@google.com>
Date:   Thu Oct 19 21:36:40 2017 +0000

    Merge "Use -Werror in external/v8" am: 91ca7d58d4 am: 25eea94176
    am: 1d75d7df90

    Change-Id: I5061d0f01600922b97cf014f4627a45756ab28c7

commit 1d75d7df90bcc70961ce8a09d53966521a374699
Merge: b951f125 25eea941
Author: Chih-hung Hsieh <chh@google.com>
Date:   Thu Oct 19 21:32:49 2017 +0000

    Merge "Use -Werror in external/v8" am: 91ca7d58d4
    am: 25eea94176

    Change-Id: Ic64f1904ec11b81d84a9fd4d64d2f81a2cc2c86f

commit 25eea94176b419fc0efee84a0ab2b2b05f037448
Merge: b951f125 91ca7d58
Author: Chih-hung Hsieh <chh@google.com>
Date:   Thu Oct 19 21:30:13 2017 +0000

    Merge "Use -Werror in external/v8"
    am: 91ca7d58d4

    Change-Id: Ideedc29929962e153de1428735ee7ed64e625c7c

commit 91ca7d58d4c7ef237f047c24f23923a4153478d2
Merge: 37a30838 c6097ca5
Author: Chih-hung Hsieh <chh@google.com>
Date:   Thu Oct 19 21:18:20 2017 +0000

    Merge "Use -Werror in external/v8"

commit c6097ca5e1f0fbca0efd80a321fa8ae0a2652823
Author: Chih-Hung Hsieh <chh@google.com>
Date:   Wed Oct 11 14:37:40 2017 -0700

    Use -Werror in external/v8

    Bug: 66996870
    Test: build with WITH_TIDY=1
    Exempt-From-Owner-Approval: Android only -Wall found no new warning
    Change-Id: I1cec2abc9da9310278e77685bef68120501da67a

Change-Id: I9449e07efc9eb7800c6ca24189285c9585a7fa42
---
 Android.libv8.mk                |   2 +
 Android.v8common.mk             |   3 ++
 src/ast/ast-numbering.cc        |   2 +
 src/builtins/builtins-string.cc |  16 ++++--
 src/compiler/access-info.cc     |   3 +-
 src/compiler/typer.cc           |   2 +-
 src/crankshaft/hydrogen.cc      |   3 +-
 src/objects.cc                  | 109 ++++++++++++++++++++++++----------------
 src/objects.h                   |   3 ++
 9 files changed, 93 insertions(+), 50 deletions(-)

diff --git a/Android.libv8.mk b/Android.libv8.mk
index 465b8b4..1d96e6b 100644
--- a/Android.libv8.mk
+++ b/Android.libv8.mk
@@ -8,6 +8,8 @@ include $(CLEAR_VARS)
 LOCAL_MODULE := libv8
 LOCAL_MODULE_CLASS := STATIC_LIBRARIES
 
+LOCAL_CFLAGS := -Wall -Werror
+
 LOCAL_WHOLE_STATIC_LIBRARIES := libv8base libv8platform libv8sampler libv8src libv8gen v8peephole
 
 LOCAL_EXPORT_C_INCLUDE_DIRS := \
diff --git a/Android.v8common.mk b/Android.v8common.mk
index d00ef1d..bfe9ae0 100644
--- a/Android.v8common.mk
+++ b/Android.v8common.mk
@@ -3,10 +3,13 @@ LOCAL_CXX_STL := libc++
 LOCAL_CPP_EXTENSION := cc
 
 LOCAL_CFLAGS += \
+	-Wall \
+	-Werror \
 	-Wno-endif-labels \
 	-Wno-import \
 	-Wno-format \
 	-Wno-unused-parameter \
+	-Wno-unused-private-field \
 	-Wno-sign-compare \
 	-Wno-missing-field-initializers \
 	-Wno-ignored-qualifiers \
diff --git a/src/ast/ast-numbering.cc b/src/ast/ast-numbering.cc
index 499760d..811bd33 100644
--- a/src/ast/ast-numbering.cc
+++ b/src/ast/ast-numbering.cc
@@ -616,6 +616,8 @@ void AstNumberingVisitor::VisitStatements(ZoneList<Statement*>* statements) {
   if (statements == NULL) return;
   for (int i = 0; i < statements->length(); i++) {
     Visit(statements->at(i));
+    if (statements->at(i)->IsJump())
+        break;
   }
 }
 
diff --git a/src/builtins/builtins-string.cc b/src/builtins/builtins-string.cc
index 7cef567..04be7a3 100644
--- a/src/builtins/builtins-string.cc
+++ b/src/builtins/builtins-string.cc
@@ -139,6 +139,7 @@ class StringBuiltinsAssembler : public CodeStubAssembler {
   typedef std::function<Node*()> NodeFunction0;
   typedef std::function<Node*(Node* fn)> NodeFunction1;
   void MaybeCallFunctionAtSymbol(Node* const context, Node* const object,
+                                 Node* const maybe_string,
                                  Handle<Symbol> symbol,
                                  const NodeFunction0& regexp_call,
                                  const NodeFunction1& generic_call);
@@ -1189,8 +1190,9 @@ void StringBuiltinsAssembler::RequireObjectCoercible(Node* const context,
 }
 
 void StringBuiltinsAssembler::MaybeCallFunctionAtSymbol(
-    Node* const context, Node* const object, Handle<Symbol> symbol,
-    const NodeFunction0& regexp_call, const NodeFunction1& generic_call) {
+    Node* const context, Node* const object, Node* const maybe_string,
+    Handle<Symbol> symbol, const NodeFunction0& regexp_call,
+    const NodeFunction1& generic_call) {
   Label out(this);
 
   // Smis definitely don't have an attached symbol.
@@ -1220,9 +1222,15 @@ void StringBuiltinsAssembler::MaybeCallFunctionAtSymbol(
   }
 
   // Take the fast path for RegExps.
+  // There's two conditions: {object} needs to be a fast regexp, and
+  // {maybe_string} must be a string (we can't call ToString on the fast path
+  // since it may mutate {object}).
   {
     Label stub_call(this), slow_lookup(this);
 
+    GotoIf(TaggedIsSmi(maybe_string), &slow_lookup);
+    GotoIfNot(IsString(maybe_string), &slow_lookup);
+
     RegExpBuiltinsAssembler regexp_asm(state());
     regexp_asm.BranchIfFastRegExp(context, object, object_map, &stub_call,
                                   &slow_lookup);
@@ -1267,7 +1275,7 @@ TF_BUILTIN(StringPrototypeReplace, StringBuiltinsAssembler) {
   // Redirect to replacer method if {search[@@replace]} is not undefined.
 
   MaybeCallFunctionAtSymbol(
-      context, search, isolate()->factory()->replace_symbol(),
+      context, search, receiver, isolate()->factory()->replace_symbol(),
       [=]() {
         Callable tostring_callable = CodeFactory::ToString(isolate());
         Node* const subject_string =
@@ -1435,7 +1443,7 @@ TF_BUILTIN(StringPrototypeSplit, StringBuiltinsAssembler) {
   // Redirect to splitter method if {separator[@@split]} is not undefined.
 
   MaybeCallFunctionAtSymbol(
-      context, separator, isolate()->factory()->split_symbol(),
+      context, separator, receiver, isolate()->factory()->split_symbol(),
       [=]() {
         Callable tostring_callable = CodeFactory::ToString(isolate());
         Node* const subject_string =
diff --git a/src/compiler/access-info.cc b/src/compiler/access-info.cc
index 8fef2f0..d563e1d 100644
--- a/src/compiler/access-info.cc
+++ b/src/compiler/access-info.cc
@@ -231,7 +231,8 @@ bool AccessInfoFactory::ComputeElementAccessInfos(
   MapTransitionList transitions(maps.length());
   for (Handle<Map> map : maps) {
     if (Map::TryUpdate(map).ToHandle(&map)) {
-      Map* transition_target =
+      Map* transition_target = map->is_stable() ?
+          nullptr :
           map->FindElementsKindTransitionedMap(&possible_transition_targets);
       if (transition_target == nullptr) {
         receiver_maps.Add(map);
diff --git a/src/compiler/typer.cc b/src/compiler/typer.cc
index ed1a04a..09003ca 100644
--- a/src/compiler/typer.cc
+++ b/src/compiler/typer.cc
@@ -1431,7 +1431,7 @@ Type* Typer::Visitor::JSCallTyper(Type* fun, Typer* t) {
           return Type::String();
         case kStringIndexOf:
         case kStringLastIndexOf:
-          return Type::Range(-1.0, String::kMaxLength - 1.0, t->zone());
+          return Type::Range(-1.0, String::kMaxLength, t->zone());
         case kStringEndsWith:
         case kStringIncludes:
           return Type::Boolean();
diff --git a/src/crankshaft/hydrogen.cc b/src/crankshaft/hydrogen.cc
index d55bb37..df7b02f 100644
--- a/src/crankshaft/hydrogen.cc
+++ b/src/crankshaft/hydrogen.cc
@@ -7176,7 +7176,8 @@ HValue* HOptimizedGraphBuilder::HandlePolymorphicElementAccess(
   // Get transition target for each map (NULL == no transition).
   for (int i = 0; i < maps->length(); ++i) {
     Handle<Map> map = maps->at(i);
-    Map* transitioned_map =
+    Map* transitioned_map = map->is_stable() ?
+        nullptr :
         map->FindElementsKindTransitionedMap(&possible_transitioned_maps);
     if (transitioned_map != nullptr) {
       transition_target.Add(handle(transitioned_map));
diff --git a/src/objects.cc b/src/objects.cc
index 9f9a628..d5b7777 100644
--- a/src/objects.cc
+++ b/src/objects.cc
@@ -12670,6 +12670,53 @@ void JSFunction::EnsureHasInitialMap(Handle<JSFunction> function) {
   map->StartInobjectSlackTracking();
 }
 
+namespace {
+bool FastInitializeDerivedMap(Isolate* isolate, Handle<JSFunction> new_target,
+                              Handle<JSFunction> constructor,
+                              Handle<Map> constructor_initial_map) {
+  // Check that |function|'s initial map still in sync with the |constructor|,
+  // otherwise we must create a new initial map for |function|.
+  if (new_target->has_initial_map() &&
+      new_target->initial_map()->GetConstructor() == *constructor) {
+    DCHECK(new_target->instance_prototype()->IsJSReceiver());
+    return true;
+  }
+  // Create a new map with the size and number of in-object properties
+  // suggested by |function|.
+
+  // Link initial map and constructor function if the new.target is actually a
+  // subclass constructor.
+  if (!IsDerivedConstructor(new_target->shared()->kind())) return false;
+
+  Handle<Object> prototype(new_target->instance_prototype(), isolate);
+  InstanceType instance_type = constructor_initial_map->instance_type();
+  DCHECK(CanSubclassHaveInobjectProperties(instance_type));
+
+  int internal_fields =
+      JSObject::GetInternalFieldCount(*constructor_initial_map);
+  int pre_allocated = constructor_initial_map->GetInObjectProperties() -
+                      constructor_initial_map->unused_property_fields();
+  int instance_size;
+  int in_object_properties;
+  new_target->CalculateInstanceSizeForDerivedClass(
+      instance_type, internal_fields, &instance_size,
+      &in_object_properties);
+
+  int unused_property_fields = in_object_properties - pre_allocated;
+  Handle<Map> map =
+      Map::CopyInitialMap(constructor_initial_map, instance_size,
+                          in_object_properties, unused_property_fields);
+  map->set_new_target_is_base(false);
+
+  JSFunction::SetInitialMap(new_target, map, prototype);
+  map->SetConstructor(*constructor);
+  map->set_construction_counter(Map::kNoSlackTracking);
+  map->StartInobjectSlackTracking();
+
+  return true;
+}
+
+}  // namespace
 
 // static
 MaybeHandle<Map> JSFunction::GetDerivedMap(Isolate* isolate,
@@ -12688,42 +12735,10 @@ MaybeHandle<Map> JSFunction::GetDerivedMap(Isolate* isolate,
 
     // Check that |function|'s initial map still in sync with the |constructor|,
     // otherwise we must create a new initial map for |function|.
-    if (function->has_initial_map() &&
-        function->initial_map()->GetConstructor() == *constructor) {
+    if (FastInitializeDerivedMap(isolate, function, constructor,
+                                 constructor_initial_map)) {
       return handle(function->initial_map(), isolate);
     }
-
-    // Create a new map with the size and number of in-object properties
-    // suggested by |function|.
-
-    // Link initial map and constructor function if the new.target is actually a
-    // subclass constructor.
-    if (IsDerivedConstructor(function->shared()->kind())) {
-      Handle<Object> prototype(function->instance_prototype(), isolate);
-      InstanceType instance_type = constructor_initial_map->instance_type();
-      DCHECK(CanSubclassHaveInobjectProperties(instance_type));
-      int internal_fields =
-          JSObject::GetInternalFieldCount(*constructor_initial_map);
-      int pre_allocated = constructor_initial_map->GetInObjectProperties() -
-                          constructor_initial_map->unused_property_fields();
-      int instance_size;
-      int in_object_properties;
-      function->CalculateInstanceSizeForDerivedClass(
-          instance_type, internal_fields, &instance_size,
-          &in_object_properties);
-
-      int unused_property_fields = in_object_properties - pre_allocated;
-      Handle<Map> map =
-          Map::CopyInitialMap(constructor_initial_map, instance_size,
-                              in_object_properties, unused_property_fields);
-      map->set_new_target_is_base(false);
-
-      JSFunction::SetInitialMap(function, map, prototype);
-      map->SetConstructor(*constructor);
-      map->set_construction_counter(Map::kNoSlackTracking);
-      map->StartInobjectSlackTracking();
-      return map;
-    }
   }
 
   // Slow path, new.target is either a proxy or can't cache the map.
@@ -13371,15 +13386,17 @@ void JSFunction::CalculateInstanceSizeHelper(InstanceType instance_type,
                                              int* instance_size,
                                              int* in_object_properties) {
   int header_size = JSObject::GetHeaderSize(instance_type);
-  DCHECK_LE(requested_internal_fields,
-            (JSObject::kMaxInstanceSize - header_size) >> kPointerSizeLog2);
+  int max_nof_fields =
+      (JSObject::kMaxInstanceSize - header_size) >> kPointerSizeLog2;
+  CHECK_LE(max_nof_fields, JSObject::kMaxInObjectProperties);
+  *in_object_properties = Min(requested_in_object_properties, max_nof_fields);
+  CHECK_LE(requested_internal_fields, max_nof_fields - *in_object_properties);
   *instance_size =
-      Min(header_size +
-              ((requested_internal_fields + requested_in_object_properties)
-               << kPointerSizeLog2),
-          JSObject::kMaxInstanceSize);
-  *in_object_properties = ((*instance_size - header_size) >> kPointerSizeLog2) -
-                          requested_internal_fields;
+      header_size +
+      ((requested_internal_fields + *in_object_properties) << kPointerSizeLog2);
+  CHECK_EQ(*in_object_properties,
+           ((*instance_size - header_size) >> kPointerSizeLog2) -
+               requested_internal_fields);
 }
 
 
@@ -13404,7 +13421,13 @@ void JSFunction::CalculateInstanceSizeForDerivedClass(
     if (!current->IsJSFunction()) break;
     JSFunction* func = JSFunction::cast(current);
     SharedFunctionInfo* shared = func->shared();
-    expected_nof_properties += shared->expected_nof_properties();
+    int count = shared->expected_nof_properties();
+    // Check that the estimate is sane.
+    if (expected_nof_properties <= JSObject::kMaxInObjectProperties - count) {
+      expected_nof_properties += count;
+    } else {
+      expected_nof_properties = JSObject::kMaxInObjectProperties;
+    }
     if (!IsDerivedConstructor(shared->kind())) {
       break;
     }
diff --git a/src/objects.h b/src/objects.h
index 04d3d38..289f5fc 100644
--- a/src/objects.h
+++ b/src/objects.h
@@ -2544,6 +2544,9 @@ class JSObject: public JSReceiver {
   static const int kHeaderSize = kElementsOffset + kPointerSize;
 
   STATIC_ASSERT(kHeaderSize == Internals::kJSObjectHeaderSize);
+  static const int kMaxInObjectProperties =
+      (kMaxInstanceSize - kHeaderSize) >> kPointerSizeLog2;
+  STATIC_ASSERT(kMaxInObjectProperties <= kMaxNumberOfDescriptors);
 
   typedef FlexibleBodyDescriptor<JSReceiver::kPropertiesOffset> BodyDescriptor;
 
-- 
2.7.4

